### PR TITLE
Stratum v1 mining server core

### PR DESCRIPTION
**Pull request by Boblepointu**
## Summary

Embedded Stratum v1 mining server for doged, enabling pool-less solo mining directly from the node.

This is the foundational PR — later PRs will add upstream proxy routing, AuxPoW merge-mining RPCs, and multi-chain support on top.

### What's included

| Layer | Files | Description |
|-------|-------|-------------|
| **Protocol** | `stratumprotocol.{h,cpp}` | Stratum JSON-RPC message types, line-buffered parser, serializers |
| **Config** | `stratumconfig.{h,cpp}` | `-stratum`, `-stratumbind`, `-stratumport`, `-stratumpassword`, `-stratumdiff`, `-coinbaseaddr` CLI args |
| **Worker** | `stratumworker.{h,cpp}` | Per-connection session state machine: subscribe → authorize → mining, variable difficulty retarget |
| **Job** | `stratumjob.{h,cpp}` | Job manager: polls `getblocktemplate`, splits coinbase for extranonce, builds merkle branches |
| **Submit** | `stratumsubmit.{h,cpp}` | Share validation via scrypt PoW hash, duplicate detection, difficulty check |
| **Stats** | `stratumstats.{h,cpp}` | Connection count, hashrate estimate, accepted/rejected share counters, JSON formatter |
| **Server** | `stratum.{h,cpp}` | libevent TCP listener, accept/read/dispatch loop, thread-safe worker management |
| **Lifecycle** | `init.cpp` | Start/stop stratum in the node init sequence (after chainstate is loaded) |
| **Logging** | `logging.{h,cpp}` | `BCLog::STRATUM` category |
| **Build** | `CMakeLists.txt` | Compile stratum modules into libserver, wire Boost unit test suite |
| **Unit tests** | `src/stratum/test/*.cpp` | Protocol parsing, config edge cases, worker lifecycle, job creation, share validation, stats output |
| **Functional test** | `stratum_basic.py` | End-to-end: start node with `-stratum=1`, TCP connect, subscribe, authorize |

### Architecture

```
┌─────────────────────────────────┐
│           doged node            │
│  ┌───────────────────────────┐  │
│  │     StratumServer         │  │
│  │  libevent TCP :9334       │  │
│  │  ┌─────────────────────┐  │  │
│  │  │  StratumWorker[]    │  │  │
│  │  │  subscribe/auth/    │  │  │
│  │  │  vardiff/sessions   │  │  │
│  │  └────────┬────────────┘  │  │
│  │           │               │  │
│  │  ┌────────▼────────────┐  │  │
│  │  │  StratumJobManager  │  │  │
│  │  │  getblocktemplate   │  │  │
│  │  │  coinbase split     │  │  │
│  │  │  merkle branches    │  │  │
│  │  └────────┬────────────┘  │  │
│  │           │               │  │
│  │  ┌────────▼────────────┐  │  │
│  │  │  ValidateShare()    │  │  │
│  │  │  scrypt PoW check   │  │  │
│  │  │  submitblock on hit │  │  │
│  │  └─────────────────────┘  │  │
│  └───────────────────────────┘  │
└─────────────────────────────────┘
```

### Usage

```bash
doged -stratum=1 -stratumport=9334 -coinbaseaddr=DYourDogeAddress
```

### Follow-up PRs (stacked)

- **PR2**: Upstream proxy connector + tiered routing (local → proxy → failover)
- **PR3**: AuxPoW merge-mining RPCs (`createauxblock`, `submitauxblock`)
- **PR4**: Multi-chain merge-mine manager with pluggable `IAuxChainModule`
- **PR5**: Independent per-chain target checking + AuxPoW proof construction

## Test plan

- [ ] `ninja check-stratum` — 6 Boost unit test suites pass
- [ ] `test/functional/stratum_basic.py` — e2e subscribe + authorize
- [ ] Manual: start doged with `-stratum=1`, connect with any Stratum v1 miner, verify jobs received
- [ ] Verify no regressions: `ninja check` on existing test suites

### CODE REVIEW

File: src/stratum/stratum.cpp [hunk 1]
```diff
@@ -0,0 +1,811 @@
...
+StratumServer::~StratumServer() {
+    Stop();
+}
+
+bool StratumServer::Start() {
+    m_startTime = GetTime();
+
+    // Thread-safety: BroadcastJob/BroadcastRawNotify are called from the
+    // validation thread while the event loop runs on m_eventThread.
+    evthread_use_pthreads();
```
Comment (roqqit) [inline src/stratum/stratum.cpp:85]: this will not build on WIN32 (see httpserver.cpp for example)

Comment (Boblepointu) [reply src/stratum/stratum.cpp:85]: Will check win32 compat whole PR. 😉

File: src/stratum/stratum.cpp [hunk 2]
```diff
@@ -0,0 +1,811 @@
...
+        SendResponse(session, req.id, UniValue(UniValue::VNULL), err);
+        return;
+    }
+
+    RoutingTier tier = m_router ? m_router->GetActiveTier()
+                                : RoutingTier::NONE;
+
+    // --- Proxy mode: forward upstream, response comes via callback ---
+    if (tier == RoutingTier::PROXY) {
+        // We're already under m_cs from HandleRead → ProcessMessage
```
Comment (roqqit) [inline src/stratum/stratum.cpp:423]: for sanity:
```
AssertLockHeld(m_cs);
```

File: src/init.cpp [hunk 3]
```diff
@@ -1561,6 +1568,14 @@ static bool AppInitServers(Config &config,
     }
 
     StartHTTPServer();
+
+    // Initialize Stratum mining server if enabled
+    auto stratumConfig = stratum::ParseStratumConfig(args);
+    if (stratumConfig && stratumConfig->enabled) {
+        LogPrintf("Stratum server configured on %s:%d\n",
+                  stratumConfig->bind, stratumConfig->port);
+    }
```
Comment (roqqit) [inline src/init.cpp]: parsing the config a second time has a bad code smell especially since its only purpose is one log line. can you move this to where stratum:InitStratumServer is called?

Comment (Boblepointu) [reply src/init.cpp]: Yup

File: test/functional/stratum_basic.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        return auth_response, notifications
+
+    def run_test(self):
+        node = self.nodes[0]
+
+        # Generate some blocks so the chain is active
+        self.generatetoaddress(node, 110, node.get_deterministic_priv_key().address)
+
+        # Give stratum server time to start
+        time.sleep(1)
```
Comment (roqqit) [inline test/functional/stratum_basic.py]: this will be flaky. use wait_until or check the debug log to see when the server is ready

File: src/rpc/mining.cpp [hunk 5]
```diff
@@ -1,5 +1,6 @@
 // Copyright (c) 2010 Satoshi Nakamoto
 // Copyright (c) 2009-2018 The Bitcoin Core developers
+// Copyright (c) 2025 Tobias Ruck and Alexandre Guillioud
```
Comment (roqqit) [inline src/rpc/mining.cpp]: this PR is already quite large. please remove this until you make changes to the file so we can focus on reviewing the stuff at hand

Comment (Boblepointu) [reply src/rpc/mining.cpp]: Done — removed the copyright line from this file.

File: test/functional/stratum_basic.py (hunk 4 above)
Comment (Boblepointu) [reply test/functional/stratum_basic.py]: Fixed — replaced time.sleep(1) with wait_until that checks the debug log for "Stratum: server started".

File: src/init.cpp (hunk 3 above)
Comment (Boblepointu) [reply src/init.cpp]: Done — removed the duplicate ParseStratumConfig call from AppInitServers and moved the log line to the InitStratumServer call site.

File: src/stratum/stratum.cpp (hunk 2 above)
Comment (Boblepointu) [reply src/stratum/stratum.cpp:423]: Added AssertLockHeld(m_cs) at the top of HandleSubmit.

File: src/stratum/stratum.cpp (hunk 1 above)
Comment (Boblepointu) [reply src/stratum/stratum.cpp:85]: Fixed — added the #ifdef WIN32 / evthread_use_windows_threads() / #else / evthread_use_pthreads() / #endif pattern (matching httpserver.cpp). Also replaced the POSIX-only socket headers (arpa/inet.h, netinet/in.h, sys/socket.h) with #include <compat.h> which already handles WIN32 vs POSIX portably.
